Trickle subscribe on O and G

j0sh opened the pull request:
* Orchestrator monitors incoming stream
* Gateway subscribes to outgoing stream

Orchestrator also pre-creates the incoming and outgoing streams in order to avoid race conditions.

There is currently no content for the outgoing stream unless something actually creates it, eg by manually starting an inference process. Because of that, the HTTP subscriber on the gateway side will eventually time out. 

Depends on #3211

j0sh committed:
Trickle subscribe on O and G

* Orchestrator monitors incoming stream
* Gateway subscribes to outgoing stream

Orchestrator also pre-creates the incoming and outgoing streams
in order to avoid race conditions.

leszko commented on server/ai_http.go:
```diff
@@ -271,6 +274,29 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
 
+		// Precreate the channels to avoid race conditions
+		// TODO get the expected mime type from the request
+		pubCh := trickle.NewLocalPublisher(h.trickleSrv, mid, "video/MP2T")
```
Do I understand correctly that these 4 lines are not really needed, because the channel will be anyway created when requested from Gateway. However, we create them here, just because we want to subscribe them below and we don't want to have some race condition when subscribing to the non-existing channel? Is that correct?

j0sh replied:
They are needed, because of race conditions:

* gateway may subscribe before the job sends anything into the output stream
* job may start subscribing before the gateway starts sending the input (as well as the O wanting to subscribe to the input itself)

leszko commented on server/ai_http.go:
```diff
@@ -271,6 +274,29 @@ func (h *lphttp) StartLiveVideoToVideo() http.Handler {
+					return
+				}
+				// We can do something with the segment data here
+				io.Copy(io.Discard, segment.Reader)
```
We can check the payment here, but let's say the payment check fails (so Gateway hasn't paid for the stream). How do we stop all the processing from here?

j0sh replied:
there's a few things (but I think we can still get basic payments in without waiting on these)

- channels needs to be closed on the orch side - we can use the local publisher which already has a `Close()` method or add an API directly on the server for that
- runner needs to be notified - either via end-of-stream which is sent down the subscribe leg when a channel is closed, or explicit job RPC
- we need to prevent the same channels from being recreated on the O when the next publish comes in, so need to add either a local registry + auth callback to the trickle server, or disallow implicit channel creation via POST and keep pre-creating everything on the O locally.
- and probably some end-of-stream indicator on the publish leg, to send back to the gateway

leszko commented on server/ai_live_video.go:
```diff
@@ -33,4 +36,39 @@ func startTricklePublish(url *url.URL, params aiRequestParams) {
+		slog.Info("error getting pipe for trickle-ffmpeg", "url", url, "err", err)
+	}
+
+	// read segments from trickle subscrption
```
```suggestion
	// read segments from trickle subscription
```

j0sh replied:
fixed in https://github.com/livepeer/go-livepeer/pull/3232/commits/91282eec9d4b5eda97cfb16c1996b9a53ba10e73

j0sh committed:
typo

mjh1 commented on server/ai_live_video.go:
```diff
@@ -33,4 +36,39 @@ func startTricklePublish(url *url.URL, params aiRequestParams) {
+			Fname: fmt.Sprintf("pipe:%d", r.Fd()),
+		}, []ffmpeg.TranscodeOptions{{
+			// TODO take from params
+			Oname:        "rtmp://localhost/out-stream",
```
would this be going into mediaMTX?

j0sh replied:
Yes this is hardcoded into mediamtx for now but we need to take a user configurable output destination. That is where you come in @mjh1 :)

j0sh committed:
trickle fixes

j0sh committed:
Don't re-publish our own streams

j0sh committed:
go fmt

j0sh commented:
Rebased onto master now that the ai-video merge is complete 🎉 

Also updated to latest trickle in https://github.com/livepeer/go-livepeer/pull/3232/commits/04d2b5facfa8f84a43e3ab893707b62e8d1f4933 ( + https://github.com/livepeer/go-livepeer/pull/3232/commits/1bae5551bc2184a7d206470984eed4cefbc7f841 ) to reflect changes in the [http-trickle](https://github.com/j0sh/http-trickle) repo

Plus a small bug fix https://github.com/livepeer/go-livepeer/pull/3232/commits/6e2b995640b7e532bdc90331fa735ef8daf9328f

j0sh committed:
Merge remote-tracking branch 'origin/master' into ja/ai-trickle-subscribe
